diff --git a/src/create.luau b/src/create.luau
index c035fc2..580519b 100644
--- a/src/create.luau
+++ b/src/create.luau
@@ -379,7 +379,7 @@ type Instances = {
 	Wire: Wire,
 }
 
-type function getvideproperties(instancename, instances, instancetype)
+type function getvideproperties(instancename: type, instances: type, instancetype: type)
 	local instance
 	if instancename:is("string") or instancename:is("generic") then return types.never end
 
@@ -400,7 +400,7 @@ type function getvideproperties(instancename, instances, instancetype)
 		end
 
 		for key, value in value:properties() do
-			if key:value() == "Connect" and value.read.tag == "function" then
+			if key:value() == "Connect" and value.read and value.read.tag == "function" then
 				return value.read
 			end
 		end
@@ -408,7 +408,8 @@ type function getvideproperties(instancename, instances, instancetype)
 		return false
 	end
 
-	local function insert(object)
+	local function insert(object: type)
+		if object:readparent() == nil then return end
 		for key, value in object:properties() do
 			local is_a_signal = is_signal(value.read)
 			if is_a_signal and value.read:is("table") then -- luau-lsp
@@ -494,7 +495,9 @@ type function getvideinstance(instancename, instances, instancetype)
 	return types.never
 end
 
-local c = (create :: any) ::
-	<T>(keyof<Instances> & T) -> (getvideproperties<T, Instances, Instance>) -> getvideinstance<T, Instances, Instance>
+local c = create :: 
+	((string) -> (Props) -> Instance)
+	& ((string, Props) -> Instance) -- ::
+	-- <T>(keyof<Instances> & T) -> (getvideproperties<T, Instances, Instance>) -> getvideinstance<T, Instances, Instance>
 
-return c
\ No newline at end of file
+return c
